### PR TITLE
New DataRequest

### PR DESCRIPTION
**Pull request by pgierz**
# Rework of `DataRequest`

This introduces new `DataRequest`, `DataRequestVariable`, and `DataRequestTable` classes. The old ones are copies directly from Ruby and are not very nice to work with.

Here, we implement a generic abstract base class ([`ABC`](https://docs.python.org/3/library/abc.html)) for `DataRequestVariable`. This allows us to define how a `DataRequestVariable` should look like from the `pymorize` Python side, without caring about how the information is ultimately extracted from the JSON tables. `CMIP6DataRequestVariable` is the first direct implementation used for JSON tables. This has the key feature of allowing us to be flexible when new CMIP standards inevitably come out, for example with CMIP7. 

A `DataRequestVariable` therefore *must* implement certain properties and methods to be correctly recognised, otherwise you get Python `TypeError`s when subclassing. 

# TODO List
+ [x] `DataRequestVariable` (and concrete `CMIP6DataRequestVariable`)
+ [x] `DataRequestTable` (and concrete `CMIP6DataRequestTable`, specifically `CMIP6JSONDataRequestTable` -- yes, I want to build two subclass levels here, since it *might* happen in the future that the source of the `DataRequestTable` *isn't* a JSON file. Maybe it's a HTTP request. Maybe it's a post-it note on the coffee machine. I don't know)
+ [x] `DataRequest` (maybe named `DataRequestCollection`?) -- some gathering of various `Variable`s and/or `Table`s into a bundle.
+ [x] Rework all the tests

# Still to come
+ [x] Integration of @siligam's `FlatTable` design (those parts that fit together with this)

# Copilot Summary
This pull request introduces a new abstract base class `DataRequestVariable` and its concrete implementation `CMIP6DataRequestVariable` in the `src/pymorize/data_request/variable.py` file. Additionally, it includes a unit test for the `CMIP6DataRequestVariable` class in `tests/unit/test_drv.py`.

Key changes include:

### New Class Definitions:
* `src/pymorize/data_request/variable.py`: 
  * Added `DataRequestVariable` abstract base class with various properties and methods to define a generic data request variable.
  * Added `CMIP6DataRequestVariable` class as a concrete implementation of `DataRequestVariable` for CMIP6 variables, utilizing the `dataclass` decorator for automatic method generation.
  * Implemented class methods for constructing instances from dictionaries and JSON files, and methods for converting instances to dictionary representations.

### Unit Testing:
* `tests/unit/test_drv.py`:
  * Added a test case to verify the creation of a `CMIP6DataRequestVariable` instance from a CMIP6 JSON table file.

### CODE REVIEW

**Comment by pgierz** [comment]
Tests now fail at the integration level, which is what I had expected.

**Comment by pgierz** [comment]
This pull request includes significant updates to the testing configuration, dependency management, and the `CMORizer` class in the `pymorize` module. The most important changes are grouped by theme and summarized below:

### Testing Configuration:
* Modified the pytest command in the CI workflow to include all Python files in the `tests/unit` directory. (`.github/workflows/CI-test.yaml`)
* Commented out unused imports and pytest hooks in `conftest.py` to clean up the code. (`conftest.py`) [[1]](diffhunk://#diff-a31c7ed5d35f5ed8233994868c54d625b18e6bacb6794344c4531e62bd9dde59L2-R10) [[2]](diffhunk://#diff-a31c7ed5d35f5ed8233994868c54d625b18e6bacb6794344c4531e62bd9dde59L26-R44)
* Added a warning filter to `pytest.ini` to ignore specific matplotlib-related warnings. (`pytest.ini`)

### Dependency Management:
* Added `semver` to the list of required dependencies in `setup.py`. (`setup.py`)
* Added `pytest-mock` to the list of testing dependencies in `setup.py`. (`setup.py`)

### `CMORizer` Class Enhancements:
* Introduced `_SUPPORTED_CMOR_VERSIONS` to the `CMORizer` class to specify supported CMOR versions. (`src/pymorize/cmorizer.py`)
* Added a destructor method to the `CMORizer` class to ensure the cluster is closed gracefully. (`src/pymorize/cmorizer.py`)
* Updated the `_post_init_read_bare_tables` method to use `CMIP6IgnoreTableFiles` instead of `IgnoreTableFiles`. (`src/pymorize/cmorizer.py`)
* Added validation for `cmor_version` in the `_post_init_create_data_request` method and updated the data request creation process. (`src/pymorize/cmorizer.py`)

### Data Request Module:
* Introduced a new `collection.py` file with the `DataRequest` and `CMIP6DataRequest` classes, including methods for creating data requests from tables, directories, and git repositories. (`src/pymorize/data_request/collection.py`)
* Added a new `factory.py` file to implement the `MetaFactory` metaclass and a factory creation function for managing data request subclasses. (`src/pymorize/data_request/factory.py`)

These changes collectively enhance the functionality, maintainability, and testing capabilities of the `pymorize` module.

File: src/pymorize/data_request/table.py [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+            return cls.from_dict(header)
+
+
+################################################################################
+
+
+class CMIP6DataRequestTable(DataRequestTable):
+    """DataRequestTable for CMIP6."""
+
+    # FIXME(PG): This might bite itself in the ass...
```
Comment (mandresm) [inline src/pymorize/data_request/table.py:296]: Why? I see no problem

File: src/pymorize/rule.py [hunk 2]
```diff
@@ -274,15 +276,10 @@ def expand_drvs(self):
...
-            ):
-                drv_clone.tables = [drv_table]
-                drv_clone.frequencies = [drv_freq]
-                drv_clone.cell_methods = [cell_methods]
-                drv_clone.cell_measures = [cell_measures]
-                rule_clone.data_request_variables = [drv_clone]
-                clones.append(rule_clone)
+            # FIXME: This is bad. I need to extract one rule for each table,
+            # but the newer API doesn't work as cleanly here...
+            rule_clone.data_request_variables = [drv_clone]
```
Comment (mandresm) [inline src/pymorize/rule.py:281]: I don't get it. Cannot you just loop through the tables? I'm also very confused by a data_request_variable showing up in multiple tables... but that's not a problem of the tool or this implementation, but of me not understanding CMOR

**Review by mandresm — APPROVED** [review]
Awesome @pgierz. I have a headache but it's very nice. I cannot say I understood everything but I could more or less follow. There are no many suggestions from my side, more questions, than anything. Once the suggestions are addressed feel free to merge right away.

File: src/pymorize/data_request/collection.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            Whether or not to include the table header in the variable object. Default is False.
+        """
+        self.tables = tables
+        self.variables = {}
+        for table in tables.values():
+            for variable in table.variables:
+                if flattable_variables:
+                    var_key = f"{table.table_id}.{variable.variable_id}"
+                else:
+                    var_key = variable.variable_id
```
[the remark this answers is not in the document]
Comment (pgierz) [reply src/pymorize/data_request/collection.py:77]: ....that's a problem. I need to fix that

File: src/pymorize/data_request/collection.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+        Parameters
+        ----------
+        tables : Dict[str, DataRequestTable]
+            A dictionary of tables.
+        flattable_variables: bool, optional
+            Whether or not to "flatten" tables by key, generating a unique key for each
+            variable. This is composed of the table_id and variable_id. Default is True.
+        include_table_headers_in_variables: bool, optional
+            Whether or not to include the table header in the variable object. Default is False.
```
[the remark this answers is not in the document]
Comment (pgierz) [reply src/pymorize/data_request/collection.py:68]: I think I fixed that in the other PR

File: src/pymorize/data_request/collection.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    @deprecation.deprecated(details="Use from_directory instead.")
+    @abstractmethod
+    def from_tables_dir(cls, directory: str) -> "DataRequest":
+        """Create a DataRequest from a directory of tables."""
+        raise NotImplementedError
+
+
+class CMIP6DataRequest(DataRequest):
+
+    GIT_URL = "..."
```
[the remark this answers is not in the document]
Comment (pgierz) [reply src/pymorize/data_request/collection.py:41]: The other PR has an actual implementation here.

File: src/pymorize/data_request/factory.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        if not hasattr(cls, "__abstractmethods__"):
+            for base in bases:
+                if isinstance(base, MetaFactory):
+                    # Strip the base product name from the subclass name
+                    base_name = base.__name__
+                    stripped_name = name.replace(base_name, "")
+                    base._registry[stripped_name] = cls
+
+
+def create_factory(klass):
```
[the remark this answers is not in the document]
Comment (pgierz) [reply src/pymorize/data_request/factory.py:24]: This is used in the cmorizer. I think I need to add examples here to make it more obvious

**Pull request merged**